Report Data Visualization

MImran2002 opened the pull request:
## CCE Minor Data Visualization

Fixes #
- Visualize Reports data #1303
- Creates graphical representations for the data given to the downloadable reports. Represent the information on both the reports tab and on the downloadable spreadsheets.

## Changes

- Add a new button that opens up to a modal to look at the graph
- The graph represents the student engagements for CCE Minors
- A new route was created with Ajax to retrieve data from the back-end
- Using Chart.js to produce a chart image and use html2canvas to take a snap of the chart so that it can be downloaded

## Testing

- Set up

- Checkout the branch `git checkout visualizeReportData`
- Pull the code when you are in the branch `git pull`
- In app>logic>minor.py in the getMinorProgress() function, add the follow code lines after line 118 (be mindful of indentations):
   for a in range (0, len(engagedStudentsList)-1, 2):
        engagedStudentsList[a]['hasSummer'] = "Completed"
- Reset the database `database/reset_database.sh test`
- Run the test suite to make sure all the tests are working `tests/run_tests.sh`
- Run the application `flask run`

View Chart and Download it.

- Make sure you are admin:ramsayb2
- Navigate to Minor Management on the navigation bar
- Under CCE Minor Progress there is a blue button "View Chart"
- Click the button to observe the chart
- Click the green "Download Chart" button to download an image of the chart
- Click the Engagement Bar Chart or Engagement Line Chart to see different chart

<img width="598" height="723" alt="image" src="https://github.com/user-attachments/assets/9efbe4f8-ca5e-42fc-b79e-12a2115aca8b" />
<img width="552" height="622" alt="image" src="https://github.com/user-attachments/assets/fa755d62-1424-438d-912b-ea26de755b19" />

MImran2002 committed:
report data visualization front-end

MImran2002 committed:
graph is working

MImran2002 committed:
the chart has been created and it can be downloaded

MImran2002 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into visualizeReportData

MImran2002 committed:
the chart has been created and it can be downloaded and console.log and unnecessary lines cleared up

MImran2002 committed:
Merge branch 'development' into visualizeReportData

MImran2002 committed:
add x and y labels

MImran2002 committed:
Merge branch 'visualizeReportData' of https://github.com/BCStudentSoftwareDevTeam/celts into visualizeReportData

Kafui123 commented on app/controllers/admin/minor.py:
```diff
@@ -1,11 +1,26 @@
+    if not g.current_user.isAdmin:
+        abort(403)
+    else:
+        progress_list = getMinorProgress()
```
This is Snake case. Switch it to CamelCase

MImran2002 replied:
I have resolved it

Kafui123 commented on app/controllers/admin/minor.py:
```diff
@@ -1,11 +1,26 @@
+        for progress in progress_list:
+            turnToChart.append({'name':progress["firstName"] + " " + progress["lastName"], "engagementCount" : progress['engagementCount'], "completeSummer": "Yes" if progress['hasSummer'] == "Complete" else "No"})
+        return jsonify(turnToChart)
+    
+        
+        
+    
```
Too many lines.

MImran2002 replied:
I have resolved it

MImran2002 committed:
changes are met

Kafui123 commented on app/static/js/minorAdminPage.js:
```diff
@@ -57,10 +56,91 @@ $(document).ready(function() {
+        }
+      },
+      error: function(error) {
+        console.log("error")
```
take a second look at this console.log

Kafui123 commented on app/static/js/minorAdminPage.js:
```diff
@@ -57,10 +56,91 @@ $(document).ready(function() {
+      downloadLink.click(); // stimulate the click
+    })
+  })
 
 })
 
```
Too many lines

MImran2002 committed:
changes are met again

stevensonmichel committed:
Change font and address feedback

bledsoef commented on app/static/js/minorAdminPage.js:
```diff
@@ -57,7 +56,95 @@ $(document).ready(function() {
+      data: JSON,
+      success: function(responses) {
+        for (let i = 0; i < responses.length; i++){
+          xValues.push(responses[i].engagementCount);
```
x values and y values are switched.

bledsoef commented on app/static/js/minorAdminPage.js:
```diff
@@ -57,7 +56,95 @@ $(document).ready(function() {
+                  max: maxValue,
+                  title:{
+                    display:true,
+                    text:'Hours of Engagements',
```
This should be Number of Engagements not Hours. Also probably shouldn't be half values if it can only be whole numbers.

MImran2002 committed:
change the x and y values and remove the decimals for the number of engagement.

MImran2002 committed:
Merge branch 'visualizeReportData' of https://github.com/BCStudentSoftwareDevTeam/celts into visualizeReportData

stevensonmichel committed:
Merge branch 'development' into visualizeReportData

MImran2002 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into visualizeReportData

MImran2002 commented on app/controllers/admin/minor.py:
```diff
@@ -1,11 +1,22 @@
+        abort(403)
+    else:
+        progressList = getMinorProgress()
+        turnToChart = []
```
chartinfo

BrianRamsay committed:
Merge branch 'development' into visualizeReportData

MImran2002 committed:
Merge branch 'bonnerCheckmarkProfile' of https://github.com/BCStudentSoftwareDevTeam/celts into bonnerCheckmarkProfile
orked on the username and other rawclasslevel possibility

MImran2002 committed:
Merge branch 'visualizeReportData' of https://github.com/BCStudentSoftwareDevTeam/celts into visualizeReportData

MImran2002 committed:
just fixing

MImran2002 committed:
line chart implemented

MImran2002 committed:
not working but thinking of creating a fdunction to expand the terms

MImran2002 committed:
terms population corrected, chart sizes correct and bar chart legend has been standardized

bakobagassas reviewed (commented):
This looks so cool! I loved the improvement.
I think it would be even better to have the chart as pdf rather than png

bakobagassas reviewed (commented):
Here we can see the bar char option as well as the line chart option but we are not able to change it, same for the engagement per term hover. I think it would be better to have it in a document where we can visualize both charts. Or perhaps making it in a way that when we download the bar chart, it only says Bar chart in the downloaded version and not bar chart AND line chart. 

<img width="495" height="524" alt="image" src="https://github.com/user-attachments/assets/08297857-89f0-4864-abfc-3eb4e85b99cb" />

bakobagassas reviewed (commented):
also the bar chat is a little bit confusing. It is unclear what the dependent variable is on the y-axis. Furthermore, I think it is confusing how the line graph seems to be about the number of engagements and the bar graph seems to be about the completed summer experience. We should either keep it constant by having a line graph for both engagement and summer experience OR having a bar graph and a line graph for either engagement or summer experience. 
<img width="1832" height="1033" alt="image" src="https://github.com/user-attachments/assets/2cf1acc4-6e91-45e7-a743-138bfcc7a94e" />

MImran2002 commented:
> This looks so cool! I loved the improvement. I think it would be even better to have the chart as pdf rather than png

1. I think about it too but I decided not to implement pdf because the chart isn't a stand-alone document but it is exported as png so that CELTS can used it in their presentation so a photo png will be more suitable.
2. I have removed the buttons from the images that will be downloaded.
3. I realize the image needs a title so I added them in.
4. I have also edited the button to better represent what the data want to visualize it.

MImran2002 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into visualizeReportData
I have changed so that the buttons do not get downloaded in the pictures, a new data visualization and renaming and adding x and y labels.

MImran2002 committed:
committing the new changes so when downloaded th button labels are correct

bakobagassas reviewed (commented):
The graph purpose has been standardized, that's nice! I think now we should not have "summer complete" and "summer incomplete" anymore on the bar graph since the graph is not dealing with summer completion anymore. Furthermore, I think the color of the bar graph should hence be something like green since we are only talking about engagement and not completion anymore. <img width="391" height="417" alt="image" src="https://github.com/user-attachments/assets/0279e11f-095b-4fd6-af10-9febf3ff3f69" />

MImran2002 committed:
clean up code tweak data visualization

MImran2002 commented:
> The graph purpose has been standardized, that's nice! I think now we should not have "summer complete" and "summer incomplete" anymore on the bar graph since the graph is not dealing with summer completion anymore. Furthermore, I think the color of the bar graph should hence be something like green since we are only talking about engagement and not completion anymore. <img alt="image" width="391" height="417" src="https://private-user-images.githubusercontent.com/194523576/528162561-0279e11f-095b-4fd6-af10-9febf3ff3f69.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjYwODY1MzAsIm5iZiI6MTc2NjA4NjIzMCwicGF0aCI6Ii8xOTQ1MjM1NzYvNTI4MTYyNTYxLTAyNzllMTFmLTA5NWItNGZkNi1hZjEwLTlmZWJmM2ZmM2Y2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIxOFQxOTMwMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNTZjNWRiZjEyNDI3NzllMWY3ZTk2NzMwNDUwMDI0MmM0YTkyYTg4ZTMyNTgzNDdkOTM4M2JlNDkzZjhiOTljJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.MzupxTp8kidOrIEPqCjK2VPOtHfuvTt2pFsPtgIfLGc">

Hello we talked about it I agreed we should change colors so I did and I decided to make it distinct based on summer and non summer term

<img width="470" height="408" alt="image" src="https://github.com/user-attachments/assets/f00b0a43-ea5b-4265-9e58-43a4776eef32" />

bakobagassas reviewed (approved):
Looks good to me now!
